File: 86-domain-service/domain_service_demo.py
# Ngôn ngữ: Python
# Công dụng: Bản Python của cùng bài học — domain service là chỗ đặt hành vi không
# thuộc về entity nào; ba loại "service" phải phân biệt; và hai lỗi đối xứng: mô hình
# THIẾU MÁU và nhét hành vi liên-aggregate vào một entity.
# Tại sao cần học: Python là ngôn ngữ dễ rơi vào mô hình thiếu máu nhất trong ba ngôn
# ngữ — `@dataclass` cho ra một entity toàn field public trong ba dòng, và gạch dưới
# `_so_du` không chặn được ai. Nhưng Python cũng cho công cụ phòng thủ tốt nhất:
# `@property` không kèm setter làm cho `tk.so_du = -500_000` ném lỗi NGAY, và `vars()`
# cho phép ĐO tỉ lệ thiếu máu của mô hình bằng một bài test chạy trong CI. Bài dựng
# đúng hai thứ đó.

from dataclasses import dataclass


# =====================================================================
# SAI 1 — MÔ HÌNH THIẾU MÁU: dataclass toàn field public + hàm rời
# =====================================================================
@dataclass
class TaiKhoanThieuMau:
    ma: str
    so_du: int          # <- cửa mở toang, và ở Python không cần viết setter nào


def rut_thieu_mau(tk, tien):
    """Luật "không được âm" nằm ở ĐÂY, nên nó chỉ có hiệu lực với ai đi qua đây."""
    if tk.so_du < tien:
        raise RuntimeError("không đủ số dư")
    tk.so_du -= tien


# =====================================================================
# ĐÚNG — entity giữ luật của chính nó
# =====================================================================
class TaiKhoan:
    def __init__(self, ma, so_du):
        if so_du < 0:
            raise ValueError("số dư ban đầu không âm")
        self._ma, self._so_du = ma, so_du

    # Luật nằm TRONG entity -> không có đường vòng nào.
    def rut(self, tien):
        if tien <= 0:
            raise ValueError("số tiền rút phải dương")
        if self._so_du < tien:
            raise RuntimeError("không đủ số dư")
        self._so_du -= tien

    def nap(self, tien):
        if tien <= 0:
            raise ValueError("số tiền nạp phải dương")
        self._so_du += tien

    # `property` KHÔNG kèm setter: đọc được, gán vào là lỗi. Đây là thứ gần nhất với
    # `private` mà Python có, và nó đủ để chặn đường vòng.
    @property
    def ma(self):
        return self._ma

    @property
    def so_du(self):
        return self._so_du


@dataclass(frozen=True)
class BieuPhi:
    """Bảng phí là VALUE OBJECT truyền vào, không phải repository được tiêm."""
    nguong: int
    phi_thap: int
    phi_cao: int

    def tinh_phi(self, so_tien):
        return self.phi_thap if so_tien <= self.nguong else self.phi_cao


@dataclass(frozen=True)
class BienLaiChuyenTien:
    tu_tai_khoan: str
    den_tai_khoan: str
    so_tien: int
    phi: int


# =====================================================================
# DOMAIN SERVICE — ở Python nó là một HÀM Ở MỨC MODULE, không phải một lớp
# =====================================================================
def chuyen_tien(tu: TaiKhoan, den: TaiKhoan, so_tien: int, bieu_phi: BieuPhi):
    """Chuyển tiền dính tới HAI tài khoản ngang nhau. Đặt nó vào `TaiKhoan.chuyen_toi()`
    là bắt một aggregate sửa một aggregate khác — đúng thứ bài 83 cấm. Đặt nó vào tầng
    ứng dụng thì luật tính phí (một luật NGHIỆP VỤ) rời khỏi miền.

    Hàm này không có biến toàn cục, không đọc đồng hồ, không chạm CSDL. Trạng thái bằng
    0 không phải vì kỷ luật — vì nó là một HÀM, nó không có chỗ nào để mà cất trạng thái.
    """
    if tu.ma == den.ma:
        raise ValueError("không chuyển cho chính mình")
    phi = bieu_phi.tinh_phi(so_tien)
    tu.rut(so_tien + phi)       # mỗi entity vẫn tự giữ luật của nó
    den.nap(so_tien)
    return BienLaiChuyenTien(tu.ma, den.ma, so_tien, phi)


# =====================================================================
# Ba loại "service" — thứ hay bị gộp làm một
# =====================================================================
class KhoTaiKhoan:                       # cổng (port) của MIỀN
    def tim(self, ma):
        raise NotImplementedError

    def luu(self, tk):
        raise NotImplementedError


class GuiThongBao:                       # HẠ TẦNG
    def gui(self, ma, noi_dung):
        raise NotImplementedError


class UngDungChuyenTien:
    """Ứng dụng: điều phối. Tải, gọi miền, lưu, phát thông báo. Không tính luật nào."""

    def __init__(self, kho, thong_bao):
        self.kho, self.thong_bao = kho, thong_bao
        self.so_lan_goi_mien = 0

    def thuc_hien(self, ma_tu, ma_den, so_tien):
        tu, den = self.kho.tim(ma_tu), self.kho.tim(ma_den)
        self.so_lan_goi_mien += 1
        bl = chuyen_tien(tu, den, so_tien, BieuPhi(1_000_000, 1_000, 5_000))
        self.kho.luu(tu)
        self.kho.luu(den)
        self.thong_bao.gui(ma_tu, f"đã chuyển {so_tien}")
        return bl


# ---- Self-check ----
if __name__ == "__main__":
    # ---- 1. MÔ HÌNH THIẾU MÁU: luật ở ngoài thì luật bị lách ----
    tm = TaiKhoanThieuMau("TK-01", 100_000)
    chan = False
    try:
        rut_thieu_mau(tm, 200_000)
    except RuntimeError:
        chan = True
    assert chan, "đi qua hàm dịch vụ thì luật có hiệu lực..."

    tm.so_du = -500_000      # ...và đây là đường vòng, mở sẵn cho tất cả mọi người
    assert tm.so_du == -500_000, "số dư ÂM, không ai chặn, không ngoại lệ"
    # Ở Python cái bẫy này dễ mắc nhất trong ba ngôn ngữ: `@dataclass` cho ra một entity
    # toàn field public trong ba dòng, và nó TRÔNG như một mô hình miền tử tế.

    tk = TaiKhoan("TK-01", 100_000)
    chan = False
    try:
        tk.rut(200_000)
    except RuntimeError:
        chan = True
    assert chan and tk.so_du == 100_000, "luật nằm TRONG entity -> không có đường vòng"

    khong_gan_duoc = False
    try:
        tk.so_du = -500_000
    except AttributeError:
        khong_gan_duoc = True
    assert khong_gan_duoc, "`property` không kèm setter: gán vào là lỗi NGAY"
    # Lưu ý ranh giới: `tk._so_du = -500_000` thì VẪN chạy (bài 83 phần 7). `property`
    # không phải khoá — nó là một cái cửa có biển "lối này", và nó chặn được đúng thứ cần
    # chặn: người viết code bình thường, đang vội, không cố tình phá.

    # ---- 2. PHÉP ĐO CHẠY ĐƯỢC: mô hình của bạn thiếu máu bao nhiêu phần trăm? ----
    def ti_le_hanh_vi(lop):
        """Tỉ lệ phương thức thật sự là HÀNH VI, trên tổng số thành viên công khai."""
        cong_khai = [t for t in vars(lop) if not t.startswith("_")]
        hanh_vi = [t for t in cong_khai if callable(vars(lop)[t])]
        return len(hanh_vi), len(cong_khai)

    assert ti_le_hanh_vi(TaiKhoanThieuMau)[0] == 0, "TaiKhoanThieuMau: 0 hành vi"
    hv, tong = ti_le_hanh_vi(TaiKhoan)
    assert hv == 2 and tong == 4, "TaiKhoan: 2 hành vi (rút/nạp) + 2 property chỉ đọc"
    assert not any(isinstance(vars(TaiKhoan)[t], property) and vars(TaiKhoan)[t].fset
                   for t in vars(TaiKhoan) if isinstance(vars(TaiKhoan)[t], property)), \
        "không property nào có setter — mọi thay đổi đi qua một hành vi có tên"
    # Ba dòng trên là một bài test kiến trúc chạy được: quét các lớp trong gói `mien/`,
    # fail nếu có lớp nào 0 hành vi hoặc có property kèm setter. Nó bắt đúng thời điểm
    # mô hình bắt đầu trượt về phía lược đồ CSDL đội lốt object.

    # ---- 3. KHI NÀO THÌ THẬT SỰ CẦN DOMAIN SERVICE ----
    # Ba câu hỏi, phải trả lời CÓ cả ba:
    #   (a) Hành vi này có phải LUẬT NGHIỆP VỤ không? (không phải điều phối, không phải I/O)
    #   (b) Nó có thuộc về đúng MỘT entity không? — nếu CÓ thì đặt vào entity đó, xong.
    #   (c) Ép nó vào một entity có làm entity đó phải sửa entity khác không?
    # "Chuyển tiền" trả lời: (a) có, (b) KHÔNG — hai tài khoản ngang nhau, (c) có.
    # => domain service.
    a, b = TaiKhoan("TK-A", 5_000_000), TaiKhoan("TK-B", 0)
    bl = chuyen_tien(a, b, 2_000_000, BieuPhi(1_000_000, 1_000, 5_000))
    assert bl.phi == 5_000, "trên 1 triệu -> phí cao"
    assert a.so_du == 2_995_000, "trừ cả tiền lẫn phí"
    assert b.so_du == 2_000_000, "bên nhận không chịu phí"

    # ---- 4. DOMAIN SERVICE KHÔNG CÓ CHỖ ĐỂ CẤT TRẠNG THÁI ----
    # `chuyen_tien` là một hàm mức module. Không `self`, không field, không constructor
    # để ai đó tiêm một repository vào. Muốn nó biết biểu phí thì phải TRUYỀN biểu phí.
    x, y = TaiKhoan("TK-X", 10_000), TaiKhoan("TK-Y", 0)
    assert chuyen_tien(x, y, 5_000, BieuPhi(1_000_000, 1_000, 5_000)).phi == 1_000, \
        "test domain service: hai dòng, không hạ tầng nào"
    assert chuyen_tien.__closure__ is None, "hàm không bắt biến ngoài — thật sự không trạng thái"
    # Nếu một ngày ai đó cần thêm repository vào domain service, đó là tín hiệu rõ ràng:
    # hoặc nó là application service đội lốt, hoặc dữ liệu nó cần phải được TRUYỀN VÀO.

    # ---- 5. CẠM BẪY RIÊNG CỦA PYTHON: lớp toàn `@staticmethod` ----
    class DichVuChuyenTien:          # <- KHÔNG nên viết thế này
        @staticmethod
        def chuyen(tu, den, so_tien, bieu_phi):
            return chuyen_tien(tu, den, so_tien, bieu_phi)

    assert DichVuChuyenTien.chuyen is not None, "chạy được, nhưng lớp này thừa"
    # Một lớp chỉ chứa `@staticmethod` là một namespace đội lốt object — Python đã có
    # namespace, tên nó là MODULE. Lớp thừa này gây ba hại nhỏ mà cộng lại thì lớn:
    #   1. Dụ người sau thêm `__init__(self, repo)` vì "đã là lớp rồi thì tiêm vào cho tiện";
    #   2. Làm cho hàm khó import lẻ và khó thay thế trong test;
    #   3. Nói dối về bản chất — domain service KHÔNG phải một object, nó là một phép tính.
    # Ở Java bạn buộc phải có lớp; ở Python thì không, nên đừng tự trói mình.

    # ---- 6. BA LOẠI SERVICE — bảng phân biệt ----
    #
    #                | Domain service      | Application service   | Infrastructure
    #   -------------|---------------------|-----------------------|----------------
    #   trả lời       | "luật là gì?"       | "quy trình là gì?"    | "làm thế nào?"
    #   ví dụ         | chuyen_tien()       | UngDungChuyenTien     | GuiThongBao
    #   có trạng thái | KHÔNG               | không                 | thường có
    #   chạm I/O      | KHÔNG               | có (qua interface)    | CÓ
    #   mở transaction| KHÔNG               | CÓ                    | không
    #   nằm ở tầng    | miền                | ứng dụng              | hạ tầng
    #   test cần gì   | không cần gì        | fake (bài 68)         | môi trường thật
    #
    # Sai lầm phổ biến nhất: gộp cột 1 và cột 2 thành một lớp `OrderService` dài 800
    # dòng, vừa mở transaction vừa tính luật vừa gửi email.
    class KhoBoNho(KhoTaiKhoan):
        def __init__(self):
            self.m = {}

        def tim(self, ma):
            return self.m.get(ma)

        def luu(self, tk):
            self.m[tk.ma] = tk

    class ThongBaoGia(GuiThongBao):
        def __init__(self):
            self.da_gui = []

        def gui(self, ma, noi_dung):
            self.da_gui.append(f"{ma}:{noi_dung}")

    kho, tb = KhoBoNho(), ThongBaoGia()
    kho.luu(TaiKhoan("TK-A", 5_000_000))
    kho.luu(TaiKhoan("TK-B", 0))
    ud = UngDungChuyenTien(kho, tb)
    ud.thuc_hien("TK-A", "TK-B", 500_000)
    assert kho.tim("TK-A").so_du == 4_499_000, "500.000 + phí thấp 1.000"
    assert len(tb.da_gui) == 1, "tầng ứng dụng lo thông báo — miền không biết email tồn tại"
    assert ud.so_lan_goi_mien == 1, "và nó gọi miền đúng một lần, không tự tính luật"

    # ---- 7. CẠM BẪY: `XxxService` thành thùng rác ----
    # Dấu hiệu nhận biết, theo thứ tự nặng dần:
    #   1. Tên là danh từ chung: `OrderService`, `UserManager`, `DataHandler` (bài 81).
    #   2. Nó có field là repository VÀ đồng thời chứa luật nghiệp vụ.
    #   3. Nó có phương thức thứ 15.
    #   4. Entity tương ứng chỉ còn field public (phép đo ở phần 2).
    # Domain service tốt thường là ĐÚNG MỘT hàm và tên là một ĐỘNG TỪ nghiệp vụ:
    # `chuyen_tien`, `tinh_lai_suat`, `kiem_tra_trung_lap`.

    # ---- 8. RANH GIỚI: khi nào KHÔNG cần domain service ----
    # Cám dỗ ngược lại cũng có thật: tạo `rut_tien(tk, tien)` cho việc `tk.rut(tien)`.
    # Câu hỏi (b) ở phần 3 trả lời CÓ — hành vi thuộc về đúng một entity — nên nó phải
    # nằm trong entity, và một hàm rời ở đây chỉ thêm một lớp vô nghĩa.
    #
    # Quy tắc: domain service là NGOẠI LỆ, không phải mặc định. Nếu miền của bạn có
    # nhiều service hơn entity, thì bạn đang viết mô hình thiếu máu và gọi nó là DDD.

    print("OK")

File: 86-domain-service/DomainServiceDemo.java
/*
 * Ngôn ngữ: Java
 * Công dụng: Domain service — chỗ đặt hành vi KHÔNG thuộc về entity nào. Bài phân biệt
 * ba loại "service" hay bị gộp làm một (miền / ứng dụng / hạ tầng), rồi cho nổ hai lỗi
 * đối xứng: mô hình THIẾU MÁU (entity chỉ có getter/setter, mọi luật nằm ngoài nên bất
 * biến bị lách dễ dàng) và ngược lại — nhét hành vi liên-aggregate vào một entity, khiến
 * nó phải sửa aggregate khác.
 * Tại sao cần học: "domain service" là khái niệm bị lạm dụng nhất trong DDD. Đặt tên một
 * lớp là `XxxService` rồi đổ mọi thứ vào đó là con đường ngắn nhất tới mô hình thiếu máu
 * — nơi entity thành cấu trúc dữ liệu và toàn bộ nghiệp vụ nằm trong các hàm tĩnh. Bài
 * này đưa ra ba câu hỏi lọc, và một phép đếm để biết mình đã rơi vào bẫy chưa.
 */
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class DomainServiceDemo {

    // =====================================================================
    // SAI 1 — MÔ HÌNH THIẾU MÁU: entity chỉ là túi dữ liệu
    // =====================================================================
    static final class TaiKhoanThieuMau {
        private String ma;
        private long soDu;
        TaiKhoanThieuMau(String ma, long soDu) { this.ma = ma; this.soDu = soDu; }
        String getMa() { return ma; }
        long getSoDu() { return soDu; }
        void setSoDu(long soDu) { this.soDu = soDu; }   // <- cửa mở toang
    }

    static final class TaiKhoanService {
        // Luật "không được âm" nằm ở ĐÂY, nên nó chỉ có hiệu lực với ai đi qua đây.
        static void rut(TaiKhoanThieuMau tk, long tien) {
            if (tk.getSoDu() < tien) throw new IllegalStateException("không đủ số dư");
            tk.setSoDu(tk.getSoDu() - tien);
        }
    }

    // =====================================================================
    // ĐÚNG — entity giữ luật của chính nó
    // =====================================================================
    static final class TaiKhoan {
        private final String ma;
        private long soDu;
        TaiKhoan(String ma, long soDu) {
            if (soDu < 0) throw new IllegalArgumentException("số dư ban đầu không âm");
            this.ma = ma; this.soDu = soDu;
        }
        // Luật nằm TRONG entity -> không có đường vòng nào.
        void rut(long tien) {
            if (tien <= 0) throw new IllegalArgumentException("số tiền rút phải dương");
            if (soDu < tien) throw new IllegalStateException("không đủ số dư");
            soDu -= tien;
        }
        void nap(long tien) {
            if (tien <= 0) throw new IllegalArgumentException("số tiền nạp phải dương");
            soDu += tien;
        }
        String ma() { return ma; }
        long soDu() { return soDu; }
    }

    // =====================================================================
    // DOMAIN SERVICE — hành vi thuộc về MIỀN nhưng không thuộc về entity nào
    // =====================================================================
    record BienLaiChuyenTien(String tuTaiKhoan, String denTaiKhoan, long soTien, long phi) { }

    /**
     * Chuyển tiền dính tới HAI tài khoản ngang nhau. Đặt nó vào `TaiKhoan.chuyenToi(khac)`
     * là bắt một aggregate sửa một aggregate khác — đúng thứ bài 83 cấm. Đặt nó vào tầng
     * ứng dụng thì luật tính phí (một luật NGHIỆP VỤ) rời khỏi miền.
     * Nên nó là domain service: thuộc miền, KHÔNG có trạng thái, KHÔNG chạm I/O.
     */
    static final class DichVuChuyenTien {
        // KHÔNG có field. Không repository, không đồng hồ, không cấu hình. Mọi thứ cần
        // biết đều đi vào qua tham số — đó là điều làm nó test được không cần gì cả.
        BienLaiChuyenTien chuyen(TaiKhoan tu, TaiKhoan den, long soTien, BieuPhi bieuPhi) {
            if (tu.ma().equals(den.ma()))
                throw new IllegalArgumentException("không chuyển cho chính mình");
            long phi = bieuPhi.tinhPhi(soTien);
            tu.rut(soTien + phi);      // mỗi entity vẫn tự giữ luật của nó
            den.nap(soTien);
            return new BienLaiChuyenTien(tu.ma(), den.ma(), soTien, phi);
        }
    }

    // Bảng phí là VALUE OBJECT truyền vào, không phải repository được tiêm.
    record BieuPhi(long nguong, long phiThap, long phiCao) {
        long tinhPhi(long soTien) { return soTien <= nguong ? phiThap : phiCao; }
    }

    // =====================================================================
    // Ba loại "service" — thứ hay bị gộp làm một
    // =====================================================================
    interface KhoTaiKhoan { TaiKhoan tim(String ma); void luu(TaiKhoan tk); }
    interface GuiThongBao { void gui(String maTaiKhoan, String noiDung); }   // HẠ TẦNG

    /** Ứng dụng: điều phối. Mở/đóng transaction, tải, gọi miền, lưu, phát thông báo. */
    static final class UngDungChuyenTien {
        private final KhoTaiKhoan kho;
        private final GuiThongBao thongBao;
        private final DichVuChuyenTien mien = new DichVuChuyenTien();
        int soLanGoiMien = 0;

        UngDungChuyenTien(KhoTaiKhoan kho, GuiThongBao thongBao) {
            this.kho = kho; this.thongBao = thongBao;
        }

        BienLaiChuyenTien thucHien(String maTu, String maDen, long soTien) {
            TaiKhoan tu = kho.tim(maTu);
            TaiKhoan den = kho.tim(maDen);
            soLanGoiMien++;
            BienLaiChuyenTien bl = mien.chuyen(tu, den, soTien, new BieuPhi(1_000_000, 1_000, 5_000));
            kho.luu(tu);
            kho.luu(den);
            thongBao.gui(maTu, "đã chuyển " + soTien);
            return bl;
        }
    }

    // ---- Self-check ----
    private static void check(boolean ok, String msg) {
        if (!ok) throw new AssertionError(msg);
    }

    public static void main(String[] args) {
        // ---- 1. MÔ HÌNH THIẾU MÁU: luật ở ngoài thì luật bị lách ----
        TaiKhoanThieuMau tm = new TaiKhoanThieuMau("TK-01", 100_000);
        boolean chan = false;
        try { TaiKhoanService.rut(tm, 200_000); } catch (IllegalStateException e) { chan = true; }
        check(chan, "đi qua service thì luật có hiệu lực...");

        tm.setSoDu(-500_000);      // ...và đây là đường vòng, mở sẵn cho tất cả mọi người
        check(tm.getSoDu() == -500_000, "số dư ÂM, không ai chặn, không ngoại lệ");
        // Đây là dấu hiệu nhận biết mô hình thiếu máu: mọi bất biến đều nằm ngoài entity,
        // nên chúng chỉ đúng với đoạn code nhớ gọi đúng chỗ. Một `setSoDu` là đủ để
        // toàn bộ luật nghiệp vụ trở thành lời khuyên.

        TaiKhoan tk = new TaiKhoan("TK-01", 100_000);
        chan = false;
        try { tk.rut(200_000); } catch (IllegalStateException e) { chan = true; }
        check(chan && tk.soDu() == 100_000, "luật nằm TRONG entity -> không có đường vòng");
        // Và không tồn tại `setSoDu` để mà lách: cách DUY NHẤT đổi số dư là `rut`/`nap`.

        // ---- 2. PHÉP ĐẾM: bạn đã rơi vào mô hình thiếu máu chưa? ----
        long pmThieuMau = java.util.Arrays.stream(TaiKhoanThieuMau.class.getDeclaredMethods())
                .filter(m -> m.getName().startsWith("get") || m.getName().startsWith("set")).count();
        long tongTM = TaiKhoanThieuMau.class.getDeclaredMethods().length;
        check(pmThieuMau == tongTM, "TaiKhoanThieuMau: 100% phương thức là getter/setter");

        long pmDung = java.util.Arrays.stream(TaiKhoan.class.getDeclaredMethods())
                .filter(m -> m.getName().startsWith("get") || m.getName().startsWith("set")).count();
        check(pmDung == 0, "TaiKhoan: 0 getter/setter — mọi phương thức là một HÀNH VI");
        // Phép đo chạy được: tỉ lệ getter/setter trên tổng số phương thức của các lớp
        // miền. Gần 100% nghĩa là miền của bạn là một lược đồ CSDL đội lốt object.

        // ---- 3. KHI NÀO THÌ THẬT SỰ CẦN DOMAIN SERVICE ----
        // Ba câu hỏi, phải trả lời CÓ cả ba:
        //   (a) Hành vi này có phải LUẬT NGHIỆP VỤ không? (không phải điều phối, không phải I/O)
        //   (b) Nó có thuộc về đúng MỘT entity không? — nếu CÓ thì đặt vào entity đó, xong.
        //   (c) Ép nó vào một entity có làm entity đó phải sửa entity khác không?
        // "Chuyển tiền" trả lời: (a) có, (b) KHÔNG — hai tài khoản ngang nhau, (c) có.
        // => domain service.
        TaiKhoan a = new TaiKhoan("TK-A", 5_000_000);
        TaiKhoan b = new TaiKhoan("TK-B", 0);
        BienLaiChuyenTien bl = new DichVuChuyenTien().chuyen(a, b, 2_000_000, new BieuPhi(1_000_000, 1_000, 5_000));
        check(bl.phi() == 5_000, "trên 1 triệu -> phí cao");
        check(a.soDu() == 2_995_000, "trừ cả tiền lẫn phí");
        check(b.soDu() == 2_000_000, "bên nhận không chịu phí");

        // Nếu nhét vào entity: `a.chuyenToi(b, 2_000_000)` thì `TaiKhoan` phải gọi
        // `b.nap(...)` — một aggregate sửa một aggregate khác trong cùng lời gọi, đúng
        // thứ bài 83 cấm. Và câu hỏi "phí do bên nào chịu" bỗng thành trách nhiệm của
        // lớp `TaiKhoan`, dù nó là luật của DỊCH VỤ CHUYỂN TIỀN chứ không phải của tài khoản.

        // ---- 4. DOMAIN SERVICE KHÔNG CÓ TRẠNG THÁI, KHÔNG CHẠM I/O ----
        check(DichVuChuyenTien.class.getDeclaredFields().length == 0,
                "domain service không có field nào — không kho, không đồng hồ, không cấu hình");
        // Hệ quả trực tiếp: nó test được không cần gì cả. Không mock, không fake, không
        // CSDL. Nếu domain service của bạn cần một repository để chạy, thì hoặc nó là
        // application service đội lốt, hoặc dữ liệu nó cần phải được TRUYỀN VÀO (như
        // `BieuPhi` ở trên) thay vì để nó tự đi lấy.

        // ---- 5. BA LOẠI SERVICE — bảng phân biệt ----
        //
        //                | Domain service      | Application service   | Infrastructure
        //   -------------|---------------------|-----------------------|----------------
        //   trả lời       | "luật là gì?"       | "quy trình là gì?"    | "làm thế nào?"
        //   ví dụ         | DichVuChuyenTien    | UngDungChuyenTien     | GuiThongBao
        //   có trạng thái | KHÔNG               | không                 | thường có
        //   chạm I/O      | KHÔNG               | có (qua interface)    | CÓ
        //   mở transaction| KHÔNG               | CÓ                    | không
        //   nằm ở tầng    | miền                | ứng dụng              | hạ tầng
        //   test cần gì   | không cần gì        | fake (bài 68)         | môi trường thật
        //
        // Sai lầm phổ biến nhất: gộp cột 1 và cột 2 thành một lớp `OrderService` dài 800
        // dòng, vừa mở transaction vừa tính luật vừa gửi email.
        Map<String, TaiKhoan> csdl = new LinkedHashMap<>();
        csdl.put("TK-A", new TaiKhoan("TK-A", 5_000_000));
        csdl.put("TK-B", new TaiKhoan("TK-B", 0));
        List<String> daGui = new ArrayList<>();

        UngDungChuyenTien ud = new UngDungChuyenTien(
                new KhoTaiKhoan() {
                    @Override public TaiKhoan tim(String ma) { return csdl.get(ma); }
                    @Override public void luu(TaiKhoan tk) { csdl.put(tk.ma(), tk); }
                },
                (ma, noiDung) -> daGui.add(ma + ":" + noiDung));

        ud.thucHien("TK-A", "TK-B", 500_000);
        check(csdl.get("TK-A").soDu() == 4_499_000, "500.000 + phí thấp 1.000");
        check(daGui.size() == 1, "tầng ứng dụng lo thông báo — miền không biết email tồn tại");
        check(ud.soLanGoiMien == 1, "và nó gọi miền đúng một lần, không tự tính luật");

        // ---- 6. CẠM BẪY: `XxxService` thành thùng rác ----
        // Dấu hiệu nhận biết, theo thứ tự nặng dần:
        //   1. Tên là danh từ chung: `OrderService`, `UserManager`, `DataHandler` (bài 81).
        //   2. Nó có field là repository VÀ đồng thời chứa luật nghiệp vụ.
        //   3. Nó có phương thức thứ 15.
        //   4. Entity tương ứng chỉ còn getter/setter (phép đếm ở phần 2).
        // Domain service tốt thường có ĐÚNG MỘT phương thức công khai và tên là một
        // ĐỘNG TỪ nghiệp vụ: `DichVuChuyenTien.chuyen`, `TinhLaiSuat.cho`,
        // `KiemTraTrungLap.giua`.
        long soPmCongKhai = java.util.Arrays.stream(DichVuChuyenTien.class.getDeclaredMethods())
                .filter(m -> !m.isSynthetic()).count();
        check(soPmCongKhai == 1, "một domain service, một việc");

        // ---- 7. RANH GIỚI: khi nào KHÔNG cần domain service ----
        // Cám dỗ ngược lại cũng có thật: tạo `DichVuRutTien` cho việc `tk.rut(tien)`.
        // Câu hỏi (b) ở phần 3 trả lời CÓ — hành vi thuộc về đúng một entity — nên nó
        // phải nằm trong entity, và một service ở đây chỉ thêm một lớp vô nghĩa.
        //
        // Quy tắc: domain service là NGOẠI LỆ, không phải mặc định. Nếu miền của bạn có
        // nhiều service hơn entity, thì bạn đang viết mô hình thiếu máu và gọi nó là DDD.
        check(true, "mặc định: hành vi nằm trong entity/value object; service phải có lý do");

        System.out.println("OK");
    }
}
